common/determined_common/_logging.py

```diff
@@ -1,16 +1,207 @@
+        self.quitting = True
+        with self.cond:
+            self.cond.notify_all()
+
+
+def read_level(level: Optional[str], default: int, limit: int = 0) -> int:
+    return max(logging._checkLevel(level if level is not None else default), limit)  # type: ignore
+
+
+class DebugConfig:
+    def __init__(
+        self,
+        root_log_level: Optional[str] = None,
+        storage_log_level: Optional[str] = None,
+        harness_log_level: Optional[str] = None,
+        debug_all_workers: bool = False,
+        horovod_verbose: bool = False,
+        nccl_debug: Optional[str] = None,
+        nccl_debug_subsys: Optional[str] = None,
+        resource_profile_period_sec: float = 0.0,
+        stack_trace_period_sec: float = 0.0,
+    ):
+        self._root_log_level = root_log_level
+        self._debug_all_workers = debug_all_workers
+        self._horovod_verbose = horovod_verbose
+        self._storage_log_level = storage_log_level
+        self._harness_log_level = harness_log_level
+        self._nccl_debug = nccl_debug
+        self._nccl_debug_subsys = nccl_debug_subsys
+        self._resource_profile_period_sec = resource_profile_period_sec
+        self._stack_trace_period_sec = stack_trace_period_sec
+
+        if 0.0 < self._resource_profile_period_sec < 0.1:
+            logging.warning(
+                "resource profiling with psutil and a resource_profile_period_sec less than .1 "
+                "seconds is unlikely to yield accurate results, adjusting to .1 seconds."
+            )
+            self._resource_profile_period_sec = 0.1
+
+        # The current workload, tracked here so we can emit it as part of the profiling output.
+        self._workload = ""
+
+    @classmethod
+    def from_config(cls, config: Union[None, bool, Dict[str, Any]]) -> "DebugConfig":
+        if config is None or config is False:
+            return cls()
+
+        if config is True:
+            return cls(
+                root_log_level="DEBUG",
+                storage_log_level="INFO",
+                harness_log_level="DEBUG",
+                debug_all_workers=True,
+                horovod_verbose=True,
+                nccl_debug=None,
+                nccl_debug_subsys=None,
+                resource_profile_period_sec=10.0,
+                stack_trace_period_sec=0.0,
+            )
+
+        assert isinstance(config, dict)
+        return cls(**config)
+
+    @classmethod
+    def from_environ(cls, environ: Dict[str, str]) -> "DebugConfig":
+        if environ.get("DET_DEBUG"):
+            # If DET_DEBUG is set to a non-empty string, behave like `debug: true` in the config.
+            return cls.from_config(True)
+        return cls()
+
+    def set_workload(self, workload: str) -> None:
+        """Update the workload information to associate with debug messages."""
+        self._workload = workload
+
+    @contextlib.contextmanager
+    def profiling_thread(self, machine_rank: int, worker_rank: int) -> Generator:
```

Comment:
question: is this ever used?

Reply:
it is not.  I moved the _ProfilingThread to the harness package but forgot to take this out.